### PR TITLE
convert hard coded system properties into configuration classes

### PR DESCRIPTION
**Pull request by hardmettle**
This is to make .server file mandatory as part of #1346 . Added a new `EnsimeServerConfig` on the lines of `EnsimeConfig` . The `parse` method is left unimplemented and the WIP . @fommil please take a look .

### CODE REVIEW

**Comment by fommil** [comment]
thanks!

We don't want to make this mandatory.

The initial implementation should set it up with the following fallback rules, which can be achieved using the typesafe config library:

- read the legacy system properties
- read `.ensime-server` from the same folder as the `.ensime` (if it exists), which itself is passed as a system property (this is not part of the server config)
- read `~/.ensime-server`

**Comment by yyadavalli** [comment]
Good start !!

Adding to @fommil suggestions, add a XDG compliant location for the global fallback option would be good and also system properties should be the last fallback i.e.

1.  read .ensime-server from the same folder as the .ensime (if it exists), which itself is passed as a system property (this is not part of the server config)

**Comment by fommil** [comment]
Yup, that's better

**Comment by hardmettle** [comment]
@fommil sorry for  delayed progress.I have created the fallback mechanism for server file, but there are are several tests failing.I have tried fixing lot of them but I will need some assistance from here on from you. Can you please take a look when you can.

**Comment by yyadavalli** [comment]
Looks like you are missing an implicit format for `EnsimeServerConfig` also the `parse` method is undefined causing the implicit derivation to fail

**Comment by fommil** [comment]
Thanks! I'll try to take a look

File: api/src/main/scala/org/ensime/api/config.scala [hunk 1]
```diff
@@ -101,3 +95,5 @@ final case class EnsimeProject(
 ) {
   sources.foreach(f => require(f.exists, "" + f + " is required but does not exist"))
 }
+
+final case class EnsimeServerConfig(shutDownOnDisconnect: Boolean, test: Boolean, exitAfterIndex: Boolean, disableSourceMonitoring: Boolean, disableClassMonitoring: Boolean, sourceMode: Boolean, protocol: String, ENSIME_EXPERIMENTAL_H2: String = "jdbc:h2:file:", ENSIME_SKIP_JRE_INDEX: Boolean, parallelThread: Int)
```
Comment (fommil) [inline api/src/main/scala/org/ensime/api/config.scala]: a new field on a new line is much easier to read

File: core/src/it/scala/org/ensime/fixture/AnalyzerFixture.scala [hunk 2]
```diff
@@ -31,6 +32,7 @@ trait IsolatedAnalyzerFixture
     withVFS { implicit vfs =>
       withTestKit { testkit =>
         import testkit._
+        implicit val s = EnsimeConfigProtocol.parse("")
```
Comment (fommil) [inline core/src/it/scala/org/ensime/fixture/AnalyzerFixture.scala]: why are you doing this? It'll throw an exception. But there is already an implicit EnsimeConfig in scope

File: core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala [hunk 3]
```diff
@@ -22,7 +22,7 @@ trait EnsimeConfigFixture {
 
   def copyTargets: Boolean = true
 
-  def withEnsimeConfig(testCode: EnsimeConfig => Any): Any
+  def withEnsimeConfig(testCode: (EnsimeConfig) => Any): Any
```
Comment (fommil) [inline core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala]: why did you add parentheses here?

File: core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala [hunk 4]
```diff
@@ -28,17 +27,18 @@ object EnsimeConfigProtocol {
   private implicit val moduleFormat: SexpFormat[EnsimeModule] = cachedImplicit
   private implicit val configFormat: SexpFormat[EnsimeConfig] = cachedImplicit
 
-  def parse(config: String): EnsimeConfig = {
+  def parse(config: String, serverConfig: String): EnsimeConfig = {
```
Comment (fommil) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: why would you want to parse both files at the same time?

That doesn't matter. There is no need to parse anything for the serverconfig at this time, the construction of the object should be done by reading system properties at this stage, e.g. in a convenience method in `util`. After that works, we can look at creating a HOCON parser (**not** an s-expression parser).

Comment (yyadavalli) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: @hardmettle if you need the server configuration for parsing `.ensime` , construct a `EnsimeServerConfig` either by parsing it or from system properties.

Adding a `sexp` parser for `EnsimeServerConfig` should be as simple as adding an implicit `SexpFormat` for it.

Comment (fommil) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: I think we want a HOCON file though, I'm undecided. The s-expression format is difficult for a lot of editors to handle... so I'm even thinking that `.ensime` should be `.ensime.conf` in HOCON format.

Comment (yyadavalli) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: I thought we can support both. 
I started a writing the parser for HOCON about a month ago and think got stopped, parsing Lists.
Might be @hardmettle  can pick it up if he gets to it first

Comment (yyadavalli) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: We could also use something like https://github.com/ccadllc/cedi-config but that adds another scala dependency

Comment (fommil) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: @yyadavalli we don't need to write a parser, we can just use a minimal shapeless impl and typesafe config. We don't need to support sealed traits or anything at this stage just an extremely small subset of https://github.com/melrief/pureconfig

Comment (yyadavalli) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: That's what I was working on. Same confusion as before.

Comment (fommil) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: ah ok! have a look in s-express's collection formats for some hints

Comment (yyadavalli) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: Thanks, will do, just not able to find the time these days.

Comment (fommil) [inline core/src/main/scala/org/ensime/config/EnsimeConfigProtocol.scala]: I know the feeling... but I have some light at the end of the tunnel here.

**Comment by yyadavalli** [comment]
Hi @hardmettle can I pull your changes into my branch and start a new PR

**Comment by hardmettle** [comment]
@yyadavalli @fommil Sorry folks, I have been away for quite a while. Would like to restart on this again. I havent pulled the changes in a long time. Need to revisit the progress.

**Comment by yyadavalli** [comment]
@hardmettle It is still relevant, you can pick it up from where you left it. Just rebase the recent changes and we are good.

**Comment by hardmettle** [comment]
@yyadavalli Will do a rebase, thanks for the clarification.

**Comment by hardmettle** [comment]
@fommil @yyadavalli Closing this PR as it is way behind 2.0 and rebasing is turning out to be a pain. Will start afresh.
